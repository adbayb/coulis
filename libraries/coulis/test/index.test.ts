import { beforeAll, describe, expect, test } from "vitest";

import {
	createKeyframes,
	createStyles,
	createVariants,
	extract,
	globalStyles,
} from "../src";

describe("coulis", () => {
	beforeAll(() => {
		globalStyles({
			"*,*::before,*::after": {
				boxSizing: "inherit",
			},
			".globalClass+.otherGlobalClass": {
				border: "1px solid black",
				borderRadius: 4,
			},
			"@charset": '"utf-8"',
			"@font-face": {
				fontFamily: "'AliasedHelvetica'",
				src: "local(Helvetica)",
			},
			"@import":
				"url('https://fonts.googleapis.com/css?family=Open+Sans&display=swap')",
			html: {
				boxSizing: "border-box",
			},
			"html,body": {
				backgroundColor: "lightcoral",
				fontFamily: "Open Sans, AliasedHelvetica",
				margin: 0,
				padding: 0,
			},
		});
	});

	test("should generate classNames", () => {
		expect(animationName).toBe("c77e20e50");
		expect(buttonVariants({ color: "brand", size: "large" })).toBe(
			"c72f7daf8 c6a0088d1",
		);
		expect(buttonVariants({ color: "brand", size: "medium" })).toBe(
			"c72f7daf8 ccadc8a20",
		);
		expect(buttonVariants({ color: "neutral", size: "medium" })).toBe(
			"c5ea183e8 ccadc8a20",
		);
		expect(buttonVariants({ color: "accent", size: "small" })).toBe(
			"cbb3938 c2a100aa4",
		);
		expect(classNames).toBe(
			"cdd299de8 cbb3938 ce31a3799 c18a19c73 cd0c53d28",
		);
	});

	test("should extract styles given object format", () => {
		expect(extract()).toMatchSnapshot();
	});

	test("should extract styles given string format", () => {
		// eslint-disable-next-line @typescript-eslint/restrict-template-expressions
		expect(`${extract()}`).toMatchSnapshot();
	});
});

const animationName = createKeyframes({
	25: {
		transform: "scale(1.25)",
	},
	"50%": {
		transform: "scale(1.5)",
	},
	from: {
		transform: "scale(1)",
	},
	to: {
		transform: "scale(1)",
	},
});

const styles = createStyles({
	backgroundColor: {
		keys: {
			alt({ className, declaration }) {
				return `${className}[alt]{${declaration}}`;
			},
			hover({ className, declaration }) {
				return `${className}{${declaration}}`;
			},
		},
	},
	color: true,
	padding: true,
});

const buttonVariants = createVariants(styles, {
	color: {
		accent: { backgroundColor: "lightcoral" },
		brand: { backgroundColor: "lightseagreen" },
		neutral: { backgroundColor: "lightgrey" },
	},
	size: {
		large: { padding: 18 },
		medium: { padding: 12 },
		small: { padding: 6 },
	},
});

const classNames = styles({
	backgroundColor: {
		alt: "lightgray",
		base: "lightcoral",
		hover: "lightcyan",
	},
	color: "lightblue",
	padding: 10,
});
