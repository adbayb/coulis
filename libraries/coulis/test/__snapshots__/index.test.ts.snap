// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`coulis > should extract styles given object format 1`] = `
[
  {
    "attributes": {
      "data-coulis-cache": "c77e20e50,c8e4ce7ac",
      "data-coulis-scope": "global",
    },
    "content": "@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}*,*::before,*::after{box-sizing:inherit;}",
    "toString": [Function],
  },
  {
    "attributes": {
      "data-coulis-cache": "cde6daf3c,cd66da2a4,cdc6dac18,c36214926",
      "data-coulis-scope": "shorthand",
    },
    "content": ".cde6daf3c{padding:10px;}.cd66da2a4{padding:18px;}.cdc6dac18{padding:12px;}.c36214926{padding:6px;}",
    "toString": [Function],
  },
  {
    "attributes": {
      "data-coulis-cache": "c571db958,ce9403afc,cbb3db274,c3c5816c4,c5bb3b59a,cbe211eb4",
      "data-coulis-scope": "longhand",
    },
    "content": ".c571db958:hover{background-color:lightcyan;}.ce9403afc[alt]{background-color:lightgray;}.cbb3db274{background-color:lightcoral;}.c3c5816c4{color:lightblue;}.c5bb3b59a{background-color:lightseagreen;}.cbe211eb4{background-color:lightgrey;}",
    "toString": [Function],
  },
  {
    "attributes": {
      "data-coulis-cache": "c3123b398",
      "data-coulis-scope": "conditionalShorthand",
    },
    "content": "@media (min-width: 576px){.c3123b398{padding:24px;}}",
    "toString": [Function],
  },
  {
    "attributes": {
      "data-coulis-cache": "cf33f0fbe,cec510c6c,c3d523ba0,c721688f8",
      "data-coulis-scope": "conditionalLonghand",
    },
    "content": "@media (min-width: 576px){.cf33f0fbe:hover{background-color:lightcyan;}}@media (min-width: 576px){.cec510c6c[alt]{background-color:lightgray;}}@media (min-width: 576px){.c3d523ba0{background-color:lightcoral;}}@media (min-width: 576px){.c721688f8{color:lightblue;}}",
    "toString": [Function],
  },
]
`;

exports[`coulis > should extract styles given string format 1`] = `"<style data-coulis-cache="c77e20e50,c8e4ce7ac" data-coulis-scope="global">@keyframes c77e20e50{25%{transform:scale(1.25);}50%{transform:scale(1.5);}from{transform:scale(1);}to{transform:scale(1);}}*,*::before,*::after{box-sizing:inherit;}</style><style data-coulis-cache="" data-coulis-scope="shorthand"></style><style data-coulis-cache="" data-coulis-scope="longhand"></style><style data-coulis-cache="" data-coulis-scope="conditionalShorthand"></style><style data-coulis-cache="" data-coulis-scope="conditionalLonghand"></style>"`;
